src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/MessagePackHubProtocol.cs

```diff
@@ -0,0 +1,237 @@
+        public bool TryWriteMessage(HubMessage message, Stream output)
+        {
+            WriteMessage(message, output);
+            return true;
+        }
+
+        private void WriteMessage(HubMessage message, Stream output)
+        {
+            var packer = Packer.Create(output);
+            switch (message)
+            {
+                case InvocationMessage invocationMessage:
+                    WriteInvocationMessage(invocationMessage, packer, output);
+                    break;
+                case StreamItemMessage streamItemMessage:
+                    WriteStremingItemMessage(streamItemMessage, packer, output);
+                    break;
+                case CompletionMessage completionMessage:
+                    WriteCompletionMessage(completionMessage, packer, output);
+                    break;
+                default:
+                    throw new FormatException($"Unexpected message type: {message.GetType().Name}");
+            }
+        }
+
+        private static void WriteInvocationMessage(InvocationMessage invocationMessage, Packer packer, Stream output)
+        {
+            packer.Pack(InvocationMessageType);
+            packer.PackString(invocationMessage.InvocationId);
+            packer.Pack(invocationMessage.NonBlocking);
+            packer.PackString(invocationMessage.Target);
+            packer.PackObject(invocationMessage.Arguments);
+        }
+
+        private void WriteStremingItemMessage(StreamItemMessage streamItemMessage, Packer packer, Stream output)
+        {
+            packer.Pack(StreamItemMessageType);
+            packer.PackString(streamItemMessage.InvocationId);
+            packer.PackObject(streamItemMessage.Item);
+        }
+
+        private void WriteCompletionMessage(CompletionMessage completionMessage, Packer packer, Stream output)
+        {
+            packer.Pack(CompletionMessageType);
+            packer.PackString(completionMessage.InvocationId);
+            packer.PackString(completionMessage.Error);
+            if (completionMessage.Error == null)
+            {
+                packer.Pack(completionMessage.HasResult);
+                if (completionMessage.HasResult)
+                {
+                    packer.PackObject(completionMessage.Result);
+                }
+            }
+        }
+
+        private static string ReadInvocationId(Unpacker unpacker)
+        {
+            return ReadString(unpacker, "invocationId");
+        }
+
+        private static string ReadString(Unpacker unpacker, string field)
+        {
+            Exception msgPackException = null;
+            try
+            {
+                if (unpacker.ReadString(out var value))
+                {
+                    return value;
+                }
+            }
+            catch (Exception e)
+            {
+                msgPackException = e;
+            }
+
+            throw new FormatException($"Reading '{field}' as String failed.", msgPackException);
```

Comment:
The inner exception will tell the details - it is MsgPack specific. If we were at the end of the stream and could not read there won't be any exception. This basically should not happen unless the other side sent a broken message.

I might just add comment in the code.